pkg/controller/clone/common.go

```diff
@@ -0,0 +1,277 @@
+}
+
+// MustGetStorageClassForClaim returns the storageclass for a PVC
+func MustGetStorageClassForClaim(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim) (*storagev1.StorageClass, error) {
+	sc, err := GetStorageClassForClaim(ctx, c, pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	if sc == nil {
+		return nil, fmt.Errorf("no storageclass for pvc")
+	}
+
+	return sc, nil
+}
+
+// GetCompatibleVolumeSnapshotClass returns a VolumeSNapshotClass name that works for all PVCs
+func GetCompatibleVolumeSnapshotClass(ctx context.Context, c client.Client, pvcs ...*corev1.PersistentVolumeClaim) (*string, error) {
+	var storageClasses []*storagev1.StorageClass
+	for _, pvc := range pvcs {
+		sc, err := GetStorageClassForClaim(ctx, c, pvc)
+		if err != nil {
+			return nil, err
+		}
+
+		if sc == nil {
+			return nil, nil
+		}
+
+		storageClasses = append(storageClasses, sc)
+	}
+
+	volumeSnapshotClasses := &snapshotv1.VolumeSnapshotClassList{}
+	if err := c.List(ctx, volumeSnapshotClasses); err != nil {
+		if meta.IsNoMatchError(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var candidates []string
+	for _, vcs := range volumeSnapshotClasses.Items {
+		matches := true
+		for _, sc := range storageClasses {
+			if sc.Provisioner != vcs.Driver {
+				matches = false
+				break
+			}
+		}
+		if matches {
+			candidates = append(candidates, vcs.Name)
+		}
+	}
+
+	if len(candidates) > 0 {
+		sort.Strings(candidates)
+		return &candidates[0], nil
+	}
+
+	return nil, nil
+}
+
+// SameVolumeMode returns true if all pvcs have the same volume mode
+func SameVolumeMode(pvc1 *corev1.PersistentVolumeClaim, others ...*corev1.PersistentVolumeClaim) bool {
+	vm := util.ResolveVolumeMode(pvc1.Spec.VolumeMode)
+	for _, pvc := range others {
+		if util.ResolveVolumeMode(pvc.Spec.VolumeMode) != vm {
+			return false
+		}
+	}
+	return true
+}
+
+func getResource(ctx context.Context, c client.Client, namespace, name string, obj client.Object) (bool, error) {
+	obj.SetNamespace(namespace)
+	obj.SetName(name)
+
+	err := c.Get(ctx, client.ObjectKeyFromObject(obj), obj)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	return true, nil
+}
+
+func dataVolumeGetter(ctx context.Context, c client.Client) func(name, namespace string) (*cdiv1.DataVolume, error) {
+	return func(name, namespace string) (*cdiv1.DataVolume, error) {
+		obj := &cdiv1.DataVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      name,
+			},
+		}
+
+		err := c.Get(ctx, client.ObjectKeyFromObject(obj), obj)
+		if err != nil {
+			return nil, err
+		}
+
+		return obj, nil
+	}
+}
+
+func checkQuotaExceeded(r record.EventRecorder, owner client.Object, err error) {
+	if cc.ErrQuotaExceeded(err) {
+		r.Event(owner, corev1.EventTypeWarning, cc.ErrExceededQuota, err.Error())
+	}
+}
+
+func claimBoundOrWFFC(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim) (bool, error) {
```

Comment:
I think either `isClaimBoundOrWFFC` or `isPVCBoundOrWFFC` is more consistent with the rest of our code.

Reply:
yeah